### ISSUE TITLE
Embedding Layer + SGRU

### ISSUE THREAD

**Issue by kg07** [EderSantana/seya#9]
Hi,

I am trying to use the SGRU, but running into an issue while using it with an embedding layer. I have sequences of length "maxlen", and each element of the sequence is an integer denoting a word. My vocabulary size is len(chars). I want to have a binary classification for this sequence.

I first use an embedding layer to map the integers to a vector of size "embed_length". I use the output of the embedding layer to feed the SGRU. There is a dense network at the end for binary classification.

Specifically, I build the network like this:

model = Sequential()
#Embedding Layer
model.add( Embedding(input_dim = len(chars), output_dim = embed_length, input_length=maxlen) )

#SGRU Layer
model.add(SGRU(input_dim = embed_length, input_length = maxlen, output_dim = no_cells, return_sequences=False))
h1_reset = ResetRNNState(model.layers[1].h, lambda epochs, logs: epochs%4 == 0)

#Binary Classification Layer
model.add(Dense(output_dim = 1))
model.add(Activation('sigmoid'))

model.compile(loss='binary_crossentropy', optimizer=optimizer, class_mode="binary")

I am getting an error like this:
    model.add(SGRU(input_dim = embed_length, input_length = maxlen, output_dim = lstm_cells, return_sequences=False))
TypeError: __init__() takes at least 2 arguments (5 given)

Any ideas? 

PS: I installed Keras and Seya today using "pip install". I installed Theano as you suggested in thread https://github.com/EderSantana/seya/issues/5. I am not sure how to check the versions for them, but I am assuming they are the latest versions.

**Comment by EderSantana** [comment]
You forgot the `batch_size` input argument.

**Comment by kg07** [comment]
Thanks, that works, my bad. 

I have one more question. Assume I have a 100 sequences. My batch size is 5, thus 20 batches in total. If the resetRNNState has epochs parameter set to 3 (epochs%3 == 0), does that mean the states clear after 15 (3*5) sequences are seen? Or does it mean it goes through all 100 sequences 3 times, then reset the states?

Thanks.

**Comment by EderSantana** [comment]
I believe the reset we provide should be used with `on_batch_end`. People don't usually to see the entire dataset to reset. It should be something like, do 3 gradient updates (see 3 batches) and then reset. For your dataset with 20 batches, you should be reseting something like 20/3 times per epoch.

I just noticed I used the word `epochs` in the example, but that was my bad. It should read `batches` I'll fix it.

**Comment by kg07** [comment]
Thanks, that makes sense. Though this bears another question now :)

Assume the training data is a long sequence [1,2,3,4,5,....,100]. You first split this sequence into semi-redundant sequences of length "maxlen". Let's say maxlen = 3. Assume step=1 in your example code, you would have training samples like this:

[1,2,3], [2,3,4], [3,4,5], [4,5,6],...., [98,99,100]

Assume batch_size = 4. So in each batch, you would train on:
[1,2,3], [2,3,4], [3,4,5], [4,5,6]    -> batch 1
[5,6,7], [6,7,8], [7,8,9], [8,9,10]  -> batch 2
....
[95,96,97], [96,97,98], [97,98,99], [98,99,100]  -> batch 95

When you preserve the states, between which samples are the states preserved? Specifically, it looks like you would like to preserve states between [1,2,3] and [4,5,6] samples, but not between [1,2,3] and [2,3,4] samples. The latter would mean entering a sequence of [1,2,3,2,3,4]. Does that make sense?

How do you handle this?

**Comment by EderSantana** [comment]
you are right. This code was adapted from the original Keras example where they don't have statefulness. You def. want to go [1, 2, 3] and [4, 5, 6] in the next batch, in the same position of the batch.
If you can modify the code to do that, please consider submitting a PR

**Comment by kg07** [comment]
I can modify the code - but I just want to make sure I get the "state preservation" concept right. Here is my understanding, please let me know if that is true:

When you preserve states, the state is preserved between the k'th sample of batch N and k'th sample of batch (N+1). States are not preserved between k'th and (k+1)'th samples in a batch.

As an example, in the case above, I would need to set batch_size = 3 for the state preservation concept to work. In other words:
[1,2,3], [2,3,4], [3,4,5]  -> batch 1
[4,5,6],[5,6,7], [6,7,8], -> batch 2
[7,8,9], [8,9,10], [9,10,11] -> batch 3
....
[94,95,96], [95,96,97], [96,97,98] -> batch 32

One complication is, do you shuffle the order of the samples when you split into batches? Where is that done?

Thanks

**Comment by EderSantana** [comment]
if [1..100] and [a..z] are sequences and you divide in pieces of 3. each piece should be on the same position of the batch. Your batches should look like: 
batch 1
```python
[[1, 2, 3
  a, b, c]]
```
batch 2
```python
[[3, 4, 5
  d, e, f]]
```
where rows are the batch positions and columns is "time". Did you get it?
